File: mpesa-bridge/sdk/android/src/main/kotlin/com/mpesabridge/MpesaBridge.kt
package com.mpesabridge

import com.mpesabridge.models.PaymentParams
import com.mpesabridge.models.PaymentResponse
import com.mpesabridge.models.TransactionStatus
import io.socket.client.IO
import io.socket.client.Socket
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Path
import com.google.gson.Gson
import org.json.JSONObject

interface MpesaApi {
    @POST("transactions/stk-push")
    suspend fun pay(
        @Header("x-api-key") apiKey: String,
        @Body params: Map<String, Any>
    ): retrofit2.Response<ApiResponse<PaymentData>>

    @GET("transactions/{id}")
    suspend fun getStatus(
        @Header("x-api-key") apiKey: String,
        @Path("id") id: String
    ): retrofit2.Response<ApiResponse<TransactionStatus>>
}

data class ApiResponse<T>(
    val success: Boolean,
    val message: String,
    val data: T
)

data class PaymentData(
    val transactionId: String,
    val status: String
)

class MpesaBridge(
    private val apiKey: String,
    private val baseUrl: String = "https://mpesa-bridge.onrender.com/api/v1/"
) {
    private val api: MpesaApi
    private var socket: Socket? = null
    private val gson = Gson()

    init {
        val retrofit = Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        api = retrofit.create(MpesaApi::class.java)
    }

    suspend fun pay(params: PaymentParams): Result<PaymentResponse> {
        return try {
            val body = mapOf(
                "phoneNumber" to params.phone,
                "amount" to params.amount,
                "reference" to params.reference,
                "description" to (params.description ?: "Payment for ${params.reference}")
            )

            val response = api.pay(apiKey, body)
            if (response.isSuccessful && response.body() != null) {
                val data = response.body()!!.data
                Result.success(
                    PaymentResponse(
                        success = true,
                        message = response.body()!!.message,
                        transactionId = data.transactionId,
                        status = data.status
                    )
                )
            } else {
                Result.failure(Exception(response.errorBody()?.string() ?: "Payment failed"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun subscribeToUpdates(transactionId: String, callback: (TransactionStatus) -> Unit) {
        if (socket == null) {
            val socketUrl = baseUrl.replace("/api/v1/", "")
            val opts = IO.Options()
            opts.transports = arrayOf("websocket")
            socket = IO.socket(socketUrl, opts)
            socket?.connect()
        }

        socket?.emit("join-transaction", transactionId)

        socket?.on("transaction-update") { args ->
            if (args.isNotEmpty()) {
                val data = args[0] as JSONObject
                if (data.optString("transactionId") == transactionId) {
                    val status = gson.fromJson(data.toString(), TransactionStatus::class.java)
                    callback(status)
                }
            }
        }
    }

    suspend fun getStatus(transactionId: String): Result<TransactionStatus> {
        return try {
            val response = api.getStatus(apiKey, transactionId)
            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!.data)
            } else {
                Result.failure(Exception("Failed to get status"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun dispose() {
        socket?.disconnect()
        socket = null
    }
}

File: mpesa-bridge/sdk/react-native/src/MpesaBridge.ts
import axios, { AxiosInstance } from 'axios';
import { io, Socket } from 'socket.io-client';
import { MpesaConfig, PaymentParams, PaymentResponse, TransactionStatus } from './types';

export class MpesaBridge {
    private api: AxiosInstance;
    private socket: Socket | null = null;
    private config: MpesaConfig;

    constructor(config: MpesaConfig) {
        this.config = {
            baseUrl: 'https://mpesa-bridge.onrender.com/api/v1',
            ...config
        };

        this.api = axios.create({
            baseURL: this.config.baseUrl,
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': this.config.apiKey
            }
        });
    }

    /**
     * Initiate an STK Push payment
     */
    async pay(params: PaymentParams): Promise<PaymentResponse> {
        try {
            const response = await this.api.post('/transactions/stk-push', {
                phoneNumber: params.phone,
                amount: params.amount,
                reference: params.reference,
                description: params.description || `Payment for ${params.reference}`
            });

            return {
                success: true,
                message: response.data.message,
                transactionId: response.data.data.transactionId,
                status: response.data.data.status
            };
        } catch (error: any) {
            throw new Error(error.response?.data?.message || error.message || 'Payment initiation failed');
        }
    }

    /**
     * Subscribe to real-time transaction updates
     */
    subscribeToUpdates(transactionId: string, callback: (status: TransactionStatus) => void): () => void {
        // Initialize socket connection if not already active
        if (!this.socket) {
            // Connect to the root URL (remove /api/v1)
            const socketUrl = this.config.baseUrl?.replace('/api/v1', '') || '';
            this.socket = io(socketUrl, {
                transports: ['websocket'],
                autoConnect: true
            });
        }

        // Join the transaction room
        this.socket.emit('join-transaction', transactionId);

        // Listen for updates
        const listener = (data: any) => {
            if (data.transactionId === transactionId) {
                callback(data);
            }
        };

        this.socket.on('transaction-update', listener);

        // Return cleanup function
        return () => {
            if (this.socket) {
                this.socket.off('transaction-update', listener);
                // Optional: disconnect if no more listeners, but keeping it open is usually fine
            }
        };
    }

    /**
     * Manually check transaction status
     */
    async getStatus(transactionId: string): Promise<TransactionStatus> {
        try {
            const response = await this.api.get(`/transactions/${transactionId}`);
            return response.data.data;
        } catch (error: any) {
            throw new Error(error.response?.data?.message || 'Failed to get status');
        }
    }
}
